bots/github-push

```diff
@@ -0,0 +1,154 @@
+#! /usr/bin/python3
+
+# github-push -- Push the local branch to a GitHub PR, with benefits
+#
+# This tool pushes your local changes to a PR and also adds a comment
+# that summarizes what has changed.  The idea is to cement the actual
+# history of the PR in its comments, regardless of rebases, rewrites,
+# squashed fixups, etc.
+#
+# If you don't have a upstream tracking branch, you need to specify
+# the pull request number as an argument.
+#
+# If you do have a upstream tracking branch, you can omit the pull
+# request number.  The tool will find it by looking through all open
+# pull requests on GitHub and picking the one that matches the
+# tracking branch.  Once it has been found, the number is cached in
+# the Git config file to speed up subsequent invocations.
+
+import subprocess
+import argparse
+import sys
+import time
+
+from task import github
+
+def execute(*args):
+    try:
+        output = subprocess.check_output(args, universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        sys.exit(ex.returncode)
+    return output
+
+def git(*args):
+    return execute("git", *args)
+
+def git_config(name):
+    try:
+        output = subprocess.check_output([ "git", "config", name ], universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        return None
+    return output.strip()
+
+def push_and_comment(api, repo, pull_number, remote, local, push_dest, push_ref):
+    if local == remote:
+        sys.stderr.write("Nothing to push.\n")
+        return 0
+
+    diff = git("diff", remote, local)
+    old_commits_link = "https://github.com/" + repo + "/commits/" + remote
+    diff_link = "https://github.com/" + repo + "/compare/" + remote + ".." + local
```

Comment:
The existence of this github API is a really nice little gem! So the most important thing on rewriting push requests is to record all the HEADs somewhere.